### ISSUE TITLE
Product attribute set as number returns different attributes based on the value

### ISSUE THREAD

**Issue by D1mitry** [commercetools/commercetools-dotnet-core-sdk-v2#151]
In our project we have [NumberType ](https://docs.commercetools.com/api/projects/productTypes#numbertype) attribute for the product which could have floating values or integer value. During import we set it as a `NumberAttribute` float numeric value. However inside commercetools it keeps integer values without floating point, even if it was setup like a floating value (example 13.0 would be represented like a 13). When I try to get attributes for the product SDK returns different attribute types - when it is integer value it deserializes it to `LongAttribute` and when a a float value it returns `DecimalAttribute`.

**To Reproduce**
- Set Attribute of a `NumberType` with value 13.0
- Get Attribute  it will be LongAttribute
- Update same Attribute to value 13.9
- Get Attribute  it will be `DecimalAttribute`

**Expected behavior**
I would expect it will always be the same attribute type. 

**Stack information (please complete the following information):**
 - dotnet: [e.g. Core 3.1]
 - SDK: 3.0.2

**Comment by jenschude** [comment]
The backend preserves the decimal places even if a zero digit had been provided and the SDK correctly deseriales them too:

https://github.com/commercetools/commercetools-dotnet-core-sdk-v2/blob/487f143b20b15510f346a9946b7fb907bdf1a17e/commercetools.Sdk/Tests/commercetools.Api.Serialization.Tests/AttributesTests.cs#L208-L231

The issue comes from the serialization of the attributes that floating point values with a zero digit are serialized without them. See this test which is atm failing

https://github.com/commercetools/commercetools-dotnet-core-sdk-v2/blob/487f143b20b15510f346a9946b7fb907bdf1a17e/commercetools.Sdk/Tests/commercetools.Api.Serialization.Tests/AttributesTests.cs#L233-L244

We will see if there is a way to circumvent the default behavior of the System.Text.Json Serializer to strip the zero digit of decimal values

**Comment by tomasz-soltysik** [comment]
Please remember that attribute value can be provided using also
- Merchant Center: seem not preserving `.0`, can't even save if only added it
- Import API: has only [`NumberAttribute`](https://github.com/commercetools/commercetools-dotnet-core-sdk-v2/blob/master/commercetools.Sdk/commercetools.Sdk.ImportApi/Generated/Models/Productvariants/NumberAttribute.cs) class in SDK. Having only this one class seems like a proper way, as it maps 1-1 with attribute types that can be defined in your system. Those 2 separate classes in HTTP API are only confusing, as I don't know which one to use when reading commercetools attribute of type Number.

**Comment by D1mitry** [comment]
For type safe attribute values it would be better to have proper type for each attribute instead of object, e.g. `bool` Value for `BooleanAttribute`, `decimal` value for `DecimalAttribute` etc., but for this I guess `IAttribute` interface should be generic like it was in SDK v1.

**Comment by jenschude** [comment]
Long, Decimal, BoolAttribute etc extend IGenericAttribute<T> with the method GetValue() which returns the value with the correct type.

There are different possible ways to access the attributes too:

https://github.com/commercetools/commercetools-dotnet-core-sdk-v2/blob/master/commercetools.Sdk/Tests/commercetools.Api.Serialization.Tests/AttributesTests.cs

Still we could also add an extension function to return this generic attribute like this:

```
        public IGenericAttribute<TU>? ToAttribute<TU>()
        {
            if (this is IGenericAttribute<TU> t) { return t; }

            return null;
        }
```

Access then would look like this:

```
            Assert.IsType<decimal>(attributes.Get("double").ToAttribute<decimal>().GetValue());
```

**Comment by jenschude** [comment]
Sadly something like the ToAttribute method would not work correctly for Sets etc

Looking into the issue of the serialization I found out that a decimal `13.0m` will be serialized with the fraction `.0` whereas a double `13.0d` will be serialized as `13`.

To ease the dealing with the different types on deserialization a LongAttribute will get an implicit cast operator to DecimalAttribute and the extension methods would do this cast if necessary. See #153

**Comment by gms-commercetools** [comment]
@jenschude Will you let me know how we are going to proceed with this issue - this can be through slack. PS AMER had a call with the customer today and they brought up the issue with us. Thanks

**Comment by jenschude** [comment]
I just released the 3.1.0. This version adds an option to serialize attributes as decimal only and cast operators for casting from Long to DecimalAttribute and vice versa

Following code shows how to configure the client with this option:

https://github.com/commercetools/commercetools-dotnet-core-sdk-v2/blob/4ec449fbda9107ff7393a4e4d16423409991f967/commercetools.Sdk/IntegrationTests/commercetools.Api.IntegrationTests/Products/ProductsIntegrationTests.cs#L80-L92
